### ISSUE TITLE
Track the forge and reception time of ChainSync headers

### ISSUE THREAD

**Issue by mrBliss** [input-output-hk/ouroboros-network#2801]
The network team wants to know how much time there is between the forging of a block and the reception of the corresponding header via the ChainSync protocol.

Since slot to time conversions depend on the ledger state, we can't just provide a `SlotNo -> UTCTime` function. Instead, we'll need a valid `LedgerState/View` with which we can convert the slot of a header to a `RelativeTime`. If we combine that with the `RelativeTime` at which we received the header (using `SystemTime`), we can compute the time between these two events.

The ChainSync client will then expose this information about the candidates by exposing `AnchoredFragment`s of `Timed (Header blk)`.

Note that not all headers on candidate fragments will have been received over the network, e.g., when we connect to a peer with exactly the same chain as us, none of these headers will have been received from the network. In these cases, we won't be able to provide the time between the two events.

**Comment by nfrisby** [comment]
@karknu I'm picking this up from Thomas. Thanks for the prompt.

My first high-level question: Thomas's work here sofar will provide the time the header was _most recently received_. EG if our peer switches away from chain A to some chain B and then later switches to another chain C that extends A, each header H that is on both A and C will have later reception times on the candidate C than they did on the candidate A.

Is that appropriate for your use case? Fair warning: if you say "no", then I'll have to think much harder.

**Comment by karknu** [comment]
It would not be appropriate, I'd prefer a no-value to wrong value for a header we received again.

There are two use cases for this.
1. Monitoring block propagation time. This should be an instantiations conversion from `SlotNo` to `UTCTime`. This metric will give feedback to SPOs, helping them decide if it is feasible for them to run their operation with a given ISP + HW setup. #2923 is an example where this would help.
2. Ranking Peers for P2P. Peers will be ranked based on how often they are the first to present us with a new header within a moving window. This doesn't require a `SlotNo` to `UTCTime` conversion, just away to track the recv time per header per peer.

**Comment by nfrisby** [comment]
Thanks for clarifying. I'll do some studying and design and reply back, perhaps schedule a call etc.

**Comment by nfrisby** [comment]
@karknu Edsko and I discussed this today. We iterated really quickly, trying to determine both what we could easily give you and also what data seemed most meaningful. EG We could report "the first time (relative to H's forge slot) if any that we received the header H since the last connection restart" on a whole-node and/or per-peer basis. But that would require adding some new state, which we'd like to avoid as much as possible, especially since it's not obvious how to sufficiently bound the growth of that particular cache in a clearly useful and sound way.

So we iterated from there, bouncing between alternative implementations and alternative outputs (eg the above delay). We think we came up with something solid.

Would you confirm whether the below is sufficient for your Network-side needs? Thanks.

If this is too much text or too far off-base, we should have a call.

### Context

The fundamental user need: What is the probability that blocks I forge will end up on the chain?
  
Concretization to this ticket: If I receive a block too late, then I'll forge a small fork, which has a lower chance of ending up on the ultimate chain. So I want to monitor my incoming delays, so I can assess my own network/hardware configuration and/or to penalize/favor peers.

The "magic number" so to speak is the average of 20s between active slots: I need to receive the latest block, forge on it, and send my new block, all soon enough that it reaches the next leader with high probability within that average 40s window. (I haven't double-checked my logic in that last sentence, but I think the main idea is right and clear enough.)

This ticket is motivated by that; our overall goal here is to calculate a statistic or vector thereof that characterizes the incoming block propagation delays, on both a whole-node and per-peer basis. (I'm naively assuming incoming delay is correlated with outgoing delay --- I haven't considered how to separately measure outgoing delays. I suppose we could estimate round-trip delays for blocks that we forged...)

### Mechanism

I'll make our proposal and then try to justify it.

For a sliding window of the last `N` slots behind the wallclock, for each slot, we'll record the reception time of the first _tip header_ that inhabited that slot, if any.

A _tip header_ is any header that arrives as a `RollForward` message that also reports that that header is the tip of the peer's own latest selection. In other words, it's when the two `RollForward` arguments have the same hash.

https://github.com/input-output-hk/ouroboros-network/blob/60175c64673bb449ff5c500d4319c8ba6a7d8ea2/ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/Type.hs#L89-L95

Loosely, the array is `Vector (Maybe Time)` and of fixed length `N`. Thus it's crucially bounded. `N` on scale of minutes? Hours?

Remark. It's possible that the peer will extend its chain by multiple blocks in rapid succession (eg if it made a relatively long fork switch), and so we'll only identify the latest of them as a tip header. This incompleteness seems fine, unless we're never caught up to them -- we haven't considered how likely that will be under nominal circumstances. We did observe that a peer that is bulk syncing might rarely send us its tip header, but we'd actually want to penalize such a peer, right?

### Justification

We're assuming that the arrival time of a _tip header_ is an accurate indicator of that header's propagation delay. This seems uniquely true for tip headers; any other arrival of a header is not necessarily the first time we've seen it. Moreover, focusing on tip headers from slots near the wall clock gives us propagation delays of recent headers. That's what (we think) the node ultimately cares about when monitoring its network capabilities and when ranking peers.

The differences from the previous design:

  * We are not reporting on a per-header basis but instead per slot. This seems like an acceptable change, and in particular it seems to correctly handle the consequences of the peer jumping back and forth between neck-and-neck chains.
  
  * We might not report for every slot. In other words, every _tip header_ arrival is the first time that header has arrived from the peer, but not every header first arrives as a _tip header_. This incompleteness seems OK, considered this data is ultimately input to a statistic anyway.

  * Number of headers per minute is technically unbounded, while number of slots per minute is fixed by definition. This means the buffer size is determined solely by how much recent time we want to consider when computing performance statistics.

  * We could store the min and also the max propagation delay per-slot. Or also an average, etc. There is flexibility here, as long as the memory size is statically bounded.
  
  * We can maintain the N-slot array of propagation delays per-peer and merge them via piecewise `min`. So we have both node-wide (for SPO insight metrics) and per-peer statistics (for peer selection), via simple composition.
  
  * We can no longer re-use the candidate fragments themselves to communicate this delay info to the Network layer.
  
The main similarity to the previous design is that we'll still need an up-to-date `EpochInfo` in order to convert each recent slot to a current time, so we can subtract it from the recorded reception times to compute each individual propagation delay (eg the min and/or max for that non-empty slot).

**Comment by karknu** [comment]
I think that the proposal should work, but I'd like to separate  the responsibilities between consensus, network and tracing.
 How about:
 * Consensus provides a facility that can calculate the **header propagation delay** or **block propagation delay**. It should be possible to calculate the header propagation delay from chainsync's `MsgRollForward` and from tipsample's `MsgNextTip`and `MsgNextTipDone` . It should be possible to calculate the block propagation delay for blockfetch's `MsgBlock` . Consensus holds no state mapping headers, blocks or slot numbers to receive time or delay. 
 * Tracing provides a counter which tracks the **block propagation delay**. That is the block propagation time calculated at the time `addFetchedBlock` is called. Tracing keeps some state to only report the minimum delay per slot number.
 * Network tracks recent header and block propagation delay along with ΔQ metrics per peer. In case of duplicate propagation values per slot only the smallest value is kept. This structure will provide peer ranking information for the governor.

**Comment by nfrisby** [comment]
Yes, great pushback. I'm got carried away when I wrote this one:

> We can no longer re-use the candidate fragments themselves to communicate this delay info to the Network layer.

I'm unfamiliar with `TipSample` -- didn't even realize that mini protocol existed until just now.

But yes, we can at least start with the Consensus Layer's ChainSync client annotating some headers in the candidate fragments with their observed header propagation delay, specifically those that arrived as a _tip header_. I need to consider Block Fetch -- eg if it'll need the same restriction to tips.

**Comment by nfrisby** [comment]
@karknu I have a branch which ultimately provides a `Maybe Data.Time.NominalDiffTime` for each header on the fragment; it's `Just` if the header arrived as the "tip" of the peer sending it, `Nothing` otherwise, as we motivated above.

On my branch, I currently discard that when I pass the header fragments to BlockFetch, since I haven't updated BlockFetch at all. 

Do you know what the expected interface is for BlockFetch to receive these delays?

Do you want me to make this branch available so you can experiment with it now?

**Comment by karknu** [comment]
> Do you know what the expected interface is for BlockFetch to receive these delays?

I would prefer an extra function to `blockFetchClient` which looked something  something like:
```blockDelay :: HasHeader block => block -> m DiffTime```

> Do you want me to make this branch available so you can experiment with it now?

Yeah. With this data we can provide extra metrics which will be usefull for SPOs.

**Comment by nfrisby** [comment]
> I would prefer an extra function to `blockFetchClient` which looked something something like:
> `blockDelay :: HasHeader block => block -> m DiffTime`

I don't think `block -> m DiffTime` will be viable for the above proposal. EG that'd require `m` to maintain some cache that maps blocks to their reception time.

My branch currently anticipates that we'd expand `readCandidateChains` to something like `STM m (Map peer (AnchoredFragment (header, Maybe Data.Time.NominalDiffTime)))`.

https://github.com/input-output-hk/ouroboros-network/blob/59e71cb54669aac35df9747bd3037d09b8f63f6b/ouroboros-network/src/Ouroboros/Network/BlockFetch.hs#L140

> > Do you want me to make this branch available so you can experiment with it now?
> 
> Yeah. With this data we can provide extra metrics which will be usefull for SPOs.

I'll rebase it and share the branch name with you within the next 10 hours.

**Comment by karknu** [comment]
From my answer above "Consensus holds no state mapping headers, blocks or slot numbers to receive time or delay.".

I'm not interested in a consensus maintained cache that can provide recv times for blocks.

**Comment by nfrisby** [comment]
We had a call to clarify: the Network side design no longer needs header reception times. So we'll probably close this Issue as WontFix soon enough. (Note that BF can record its own block reception times.)

However, at least for the purpose of a block delay _metric_, Network still needs to convert slots to wallclock times. I've made the `nfrisby/tmp-network-layer-forge-time` branch for @karknu to try out. I'm still concerned that this is too-tight of coupling: the Consensus Team once paid an unexpected debt in order to eliminate pervasive/excessive use of this kind of conversion. Time is complicated in the presence of the HFC. The comment on the new `BlockFetchConsensusInterface` suggests why this case might be a relatively tame exception.

But, for now, please consider this branch temporary/exploratory. Hope it unblocks you for a bit more design/testing, though, in the mean-time.

**Comment by nfrisby** [comment]
@karknu -- based on my above comment, I wonder: did PRs #2988 #2995 resolve this Issue? Should we Close it? (I've mostly paged this workstream out during our intervening focus on Alonzo and neglected to write an update here back when it was all fresh.)

**Comment by karknu** [comment]
@nfrisby Yes, we have have what we needed for our peer ranking metric.

**Comment by nfrisby** [comment]
Thanks 👍    Closing.
